### PR TITLE
Scheduling

### PR DESCRIPTION
**Pull request by pavpanchekha**
This one had several critical fixes, including one critical comment (see doc).

### CODE REVIEW

File: book/scheduling.md [hunk 1]
```diff
@@ -117,13 +117,6 @@ That makes pages slower to load. We can fix this by creating tasks for
 running scripts:
 
 ``` {.python}
-class JSContext:
-    def run(self, script, code):
-        try:
-            self.interp.evaljs(code)
-        except dukpy.JSRuntimeError as e:
-            print("Script", script, "crashed", e)
-
```
Comment (pavpanchekha) [inline book/scheduling.md:126]: This code is already there, in Chapter 9.

File: book/scheduling.md [hunk 2]
```diff
@@ -287,20 +281,6 @@ thread to stop at that line of code. When more work comes in to do, such as in
...
-happen at erratic times, because the two threads are competing for the
-lock.[^try-it]
-
-[condition-variable]: https://docs.python.org/3/library/threading.html#threading.Condition
-
-[lock-class]: https://docs.python.org/3/library/threading.html#threading.Lock
-
-[^try-it]: Try removing this code and observe. The timers will become quite
-erratic.
-
```
Comment (pavpanchekha) [inline book/scheduling.md:303]: I found this confusing when reading; it makes much more sense _after_ the following code block, so I moved it.

File: book/scheduling.md [hunk 3]
```diff
@@ -393,15 +387,16 @@ Chapter 10 is not very useful and a huge performance footgun. Some
 browsers are now moving to deprecate synchronous `XMLHttpRequest`.]
 Python's `Thread` class lets us do better:
 
-    threading.Thread(target=callback).start()
+``` {.python .example}
+threading.Thread(target=callback).start()
+```
```
Comment (pavpanchekha) [inline book/scheduling.md:392]: This needs the gray background.

File: book/scheduling.md [hunk 4]
```diff
@@ -393,15 +387,16 @@ Chapter 10 is not very useful and a huge performance footgun. Some
...
     
-This code creates a new thread and then immediately returns. The
-`callback` then runs in parallel, on the new thread, while the initial
-thread continues to execute later code.
+This code creates a new thread and then continues to execute. The
+`callback` then runs in parallel, on the new thread.
 
 We'll implement asynchronous `XMLHttpRequest` calls using threads.
-Specifically, we'll have the browser start a thread, do the request
-and parse the response on that thread, and then schedule a `Task` to
```
Comment (pavpanchekha) [inline book/scheduling.md:404]: This was hard to parse.

File: book/scheduling.md [hunk 5]
```diff
@@ -489,7 +484,10 @@ class JSContext:
 
 Note that in the asynchronous case, the `XMLHttpRequest_send` method starts a
 thread and then immediately returns. That thread will run in parallel
-with the browser's main work until the request is done.
+with the browser's main work until the request is done.^[In theory two
+parallel requests could race while accessing the cookie jar; I'm not
+fixing this out of expediency but a proper implementation would have
+locks for the cookie jar.]
```
Comment (pavpanchekha) [inline book/scheduling.md:490]: I think this is the best we can really do at the moment. Hash operations might be atomic in Python but I didn't feel confident enough in this fact to put it in the footnote.

File: book/scheduling.md [hunk 6]
```diff
@@ -505,7 +503,8 @@ class JSContext:
 ```
 
 The `__runXHROnload` method just pulls the relevant object from
-`XHR_REQUESTS` and calls its `onload` function:
+`XHR_REQUESTS` and calls its `onload` function, which is the standard
+callback for asynchronous `XMLHttpRequest`s:
```
Comment (pavpanchekha) [inline book/scheduling.md:507]: Otherwise we never mention what `onload`'s role is.

File: book/scheduling.md [hunk 7]
```diff
@@ -532,11 +531,11 @@ mutations instead of page loads to update its state is called a
...
 more important.
 
-[^when-gmail]: GMail dates from April 2004, [soon after][xhr-history]
-enough browsers finished adding support for the API. The first
+[^when-gmail]: The first
 application to use `XMLHttpRequest` was [Outlook Web Access][outlook],
 in 1999, but it took a while for the API to make it into other
-browsers.
+browsers. GMail dates from April 2004, [soon after][xhr-history]
+enough browsers finished adding support for the API.
```
Comment (pavpanchekha) [inline book/scheduling.md:538]: Otherwise weirdly achronological.

File: book/scheduling.md [hunk 8]
```diff
@@ -1099,7 +1099,7 @@ class MeasureTime:
         self.file.flush()
 ```
 
-Here, the `name` argument to `start` should describe what kind of
+Here, the `name` argument to `time` should describe what kind of
```
Comment (pavpanchekha) [inline book/scheduling.md:1102]: We renamed the method and forgot to change this reference.

File: book/scheduling.md [hunk 9]
```diff
@@ -1361,15 +1361,14 @@ class Browser:
...
-Above, I needed to clear any pending tasks before loading a new page, because
+We need to clear any pending tasks before loading a new page, because
 those previous tasks are now invalid:
 
 ``` {.python}
 class TaskRunner:
     def clear_pending_tasks(self):
         self.condition.acquire(blocking=True)
         self.tasks.clear()
-        self.pending_scroll = None
```
Comment (pavpanchekha) [inline book/scheduling.md:1372]: This variable is entirely unused in this and also all other chapters.

File: book/scheduling.md [hunk 10]
```diff
@@ -1886,6 +1887,7 @@ browser thread's scroll offset is past the bottom of the page:
 ``` {.python}
 class Tab:
     def load(self, url, payload=None):
+        # ...
```
Comment (pavpanchekha) [inline book/scheduling.md:1890]: Not the only code in the function.

File: book/scheduling.md [hunk 11]
```diff
@@ -1895,8 +1897,6 @@ class Tab:
         return max(0, min(scroll, maxscroll))
 
     def run_animation_frame(self, scroll):
-        # ...
-        self.render()
```
Comment (pavpanchekha) [inline book/scheduling.md:1899]: Unchanged so I don't know why this is here.

File: book/scheduling.md [hunk 12]
```diff
@@ -328,6 +308,20 @@ class TaskRunner:
         self.condition.release()
 ```
 
+It's important to call `wait` at the end of the `run` loop if there is nothing
```
Comment (chrishtr) [inline book/scheduling.md:311]: Not critical IMO

File: book/scheduling.md [hunk 13]
```diff
@@ -393,15 +387,16 @@ Chapter 10 is not very useful and a huge performance footgun. Some
...
 
-    threading.Thread(target=callback).start()
+``` {.python .example}
+threading.Thread(target=callback).start()
+```
     
-This code creates a new thread and then immediately returns. The
-`callback` then runs in parallel, on the new thread, while the initial
-thread continues to execute later code.
+This code creates a new thread and then continues to execute. The
```
Comment (chrishtr) [inline book/scheduling.md:394]: I think the old text was better. "continues to execute" is ambiguous

File: book/scheduling.md (hunk 4 above)
Comment (chrishtr) [reply book/scheduling.md:404]: Not sure I agree, the old text seems fine to me and new text is not critical to fix either

File: book/scheduling.md (hunk 5 above)
Comment (chrishtr) [reply book/scheduling.md:490]: not critical to fix this in the printed book.

File: book/scheduling.md (hunk 6 above)
Comment (chrishtr) [reply book/scheduling.md:507]: What do you mean? I don't see why it's unclear.

File: book/scheduling.md (hunk 7 above)
Comment (chrishtr) [reply book/scheduling.md:538]: I disagree - this footnote is keying off of text about GMail so it's appropriate to mention ti first.

File: book/scheduling.md [hunk 14]
```diff
@@ -1016,8 +1015,9 @@ is really being spent rendering. It's important to always measure
 before optimizing, because the result is often surprising.
 
 To instrument our browser, let's have it output the
-[JSON][json] tracing format used by [chrome://tracing][chrome-tracing]
-in Chrome, [Firefox Profiler](https://profiler.firefox.com/) or
+[JSON][json] tracing format used by
```
Comment (chrishtr) [inline book/scheduling.md:1018]: Why make this change? Seems not significant.

Not critical in any case

File: book/scheduling.md [hunk 15]
```diff
@@ -1409,7 +1408,7 @@ class Chrome:
 Event handlers are mostly similar, except that we need to be careful
 to distinguish events that affect the browser chrome from those that
 affect the tab. For example, consider `handle_click`. If the user
-clicked on the browser chrome (meaning `e.y < self.chrome.bottom`), we can
+clicked on the browser chrome, we can
```
Comment (chrishtr) [inline book/scheduling.md:1411]: why this edit?

File: book/scheduling.md [hunk 16]
```diff
@@ -1527,8 +1526,8 @@ it was sent and call `set_needs_raster_and_draw` as needed. Because this call
 will come from another thread, we'll need to acquire a lock. Another important
 step is to not clear the `animation_timer` object until *after* the next
 commit occurs. Otherwise multiple rendering tasks could be queued at the same
-time. Finally, rename `scroll` to `active_tab_scroll` and `url` to
-`active_tab_url` to make more clear what they mean.
+time. Finally, save `scroll` in `active_tab_scroll` and `url` in
```
Comment (chrishtr) [inline book/scheduling.md:1529]: The text is describing a rename of existing fields..so it's correct as is?

File: book/scheduling.md [hunk 17]
```diff
@@ -1818,7 +1817,9 @@ but in this case we don't know the right scroll offset yet. We need
 the main thread to run in order to commit a new display list for the
 other tab, and at that point we will have a new scroll offset as well.
 Move tab switching (in `load` and `handle_click`) to a new method
-`set_active_tab` that simply schedules a new animation frame:
+`set_active_tab` that simply schedules a new animation frame:^[Note
```
Comment (chrishtr) [inline book/scheduling.md:1820]: Not critical IMO

File: book/scheduling.md (hunk 6 above)
Comment (pavpanchekha) [reply book/scheduling.md:507]: When we introduce XHR we should an example of how it's used, but that example is synchronous so it doesn't use `onload`. We never actually show an example with `onload` so someone who is not familiar with the XHR API might not understand what `onload` is or be missing the API to register an event handler.

File: book/scheduling.md (hunk 14 above)
Comment (pavpanchekha) [reply book/scheduling.md:1018]: Oh yeah, it looks weird to say "foo bar [12] in Chrome, Foo Bar [13], Foo Bar [14]". If we move "in Chrome" earlier the reference-comma becomes a consistent pattern and helps parsing because the references are big and commas are small.

Comment (pavpanchekha) [reply book/scheduling.md:1018]: Not critical.

File: book/scheduling.md (hunk 15 above)
Comment (pavpanchekha) [reply book/scheduling.md:1411]: Ah—It breaks across lines in a weird way and seems not critical so rather than rephrase text around it to avoid the weird line break I figured it's easier to delete this text. I think the relevant code used to be "e.y < 100", before we had Chrome do proper font-based layout, and back then the line of code was more cryptic, but I think now it's self-explanatory enough that we don't need to explain it.

File: book/scheduling.md (hunk 16 above)
Comment (pavpanchekha) [reply book/scheduling.md:1529]: I think it's not—it's describing a copy from the `CommitData` object into brand-new fields on the `Browser` object. Until now the `Browser` hadn't had `url` or `scroll` fields, it'd read them off the `active_tab`.

File: book/scheduling.md (hunk 4 above)
Comment (pavpanchekha) [reply book/scheduling.md:404]: Not critical, can revert, no strong feelings.

File: book/scheduling.md (hunk 12 above)
Comment (pavpanchekha) [reply book/scheduling.md:311]: I thought this was weirdly placed (it’s a move I think)

Comment (pavpanchekha) [reply book/scheduling.md:311]: But not critical.

File: book/scheduling.md (hunk 13 above)
Comment (pavpanchekha) [reply book/scheduling.md:394]: I guess my real objection is “later code” which is kind of vague. But not critical in any case.

Comment (chrishtr) [reply book/scheduling.md:394]: Is it vague? Doesn't seem to to me..

File: book/scheduling.md (hunk 4 above)
Comment (chrishtr) [reply book/scheduling.md:404]: Let's revert

File: book/scheduling.md (hunk 6 above)
Comment (chrishtr) [reply book/scheduling.md:507]: Ok. Not critical IMO

File: book/scheduling.md (hunk 13 above)
Comment (pavpanchekha) [reply book/scheduling.md:394]: If you'd rather revert it that's fine.

Comment (pavpanchekha) [reply book/scheduling.md:394]: I'll do it.

File: book/scheduling.md (hunk 15 above)
Comment (chrishtr) [reply book/scheduling.md:1412]: Not critical IMO
